### PR TITLE
Allow for specifying device with element numbers

### LINKED ISSUE #2479: Allow creation of DeviceSpecification by specifying number of elements.
### What is the expected enhancement?
Currently the `DeviceSpecification` must be initialized with a list of `Channel`s. 

https://github.com/Qiskit/qiskit-terra/blob/8b722d92221ec5988a8d4d922a57c4937c2c6d8a/qiskit/pulse/channels/device_specification.py#L27-L34

We should also provide the ability to initialize but a number of elements.

ie.,
```python
schedule = Schedule(n_qubits=5, memory_slots=5, n_control_channels, n_registers=5)
```

### PR DESCRIPTION
**Pull request by nkanazawa1989**
### Summary
close #2479 

This PR allows for specifying device with numbers of `n_drive`, `n_control`, `n_measure`, and `n_register`. As discussed in #2479, The class name `DeviceSpecification` is renamed to `PulseSpecification` because this class consists of a part of device information (mainly related to topology of device).

### Details and comments

- #qubit != `n_drive` or `n_control`
e.g. allowing for a qubit with XY control and Z control, multiple control channels, control channels to tunable coupler, etc...

- #acquire, #mem_slots = `n_measure`
i.e. these measurement chain is usually bound. This may be relevant to #2524 .

- `n_registers` != `n_measure`
e.g. measured results can be loaded to arbitrary registers and classical operations can be applied to each register.

All channels can be directly accessed without using qubit (`device.q[0].drive` and `device.d[0]` are both possible, and they refer the same object). These channels will be assigned to qubit/tunable coupler/... by using device Hamiltonian.

### CODE REVIEW

File: qiskit/pulse/channels/pulse_specification.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+class PulseSpecification:
+    """Implement a device specification."""
+
+    def __init__(self,
+                 n_drive: int,
+                 n_control: int,
+                 n_measure: int,
+                 n_registers: int,
+                 buffer: int = 0,
+                 hamiltonian=None):
```
Comment (taalexander) [inline qiskit/pulse/channels/pulse_specification.py]: This should be added as a separate PR along with all other hamiltonian related information.

File: qiskit/pulse/channels/pulse_specification.py [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+"""
+from typing import List
+
+from .channels import AcquireChannel, MemorySlot, RegisterSlot
+from .pulse_channels import DriveChannel, ControlChannel, MeasureChannel
+from .qubit import Qubit
+
+
+class PulseSpecification:
+    """Implement a device specification."""
```
Comment (taalexander) [inline qiskit/pulse/channels/pulse_specification.py]: Could you please provide a more detailed summary and description of the purpose/use case of this class?

File: qiskit/pulse/channels/pulse_specification.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+            backend (BaseBackend): Backend configuration.
+
+        Returns:
+            PulseSpecification: New PulseSpecification configured by backend.
+        """
+        configuration = backend.configuration()
+        defaults = backend.defaults()
+
+        # TODO: define n_drive and n_acquire explicitly in backend
+        n_drive = len(configuration.qubit_lo_range)
```
Comment (taalexander) [inline qiskit/pulse/channels/pulse_specification.py]: This should continue to be n_qubits, not qubit_lo_range which is guaranteed by the spec to be of length `n_qubits`. Same for measurement information. I believe you are changing this due to a future change in the spec, but this should be done as multiple steps.

File: qiskit/pulse/channels/pulse_specification.py [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+        n_measure = len(configuration.meas_lo_range)
+        n_registers = configuration.n_registers
+        buffer = defaults.buffer
+        hamiltonian = configuration.hamiltonian
+
+        return PulseSpecification(n_drive=n_drive, n_control=n_control,
+                                  n_measure=n_measure, n_registers=n_registers,
+                                  buffer=buffer, hamiltonian=hamiltonian)
+
+    def __eq__(self, other):
```
Comment (taalexander) [inline qiskit/pulse/channels/pulse_specification.py]: I'm not really sure of the use case for checking equality of two device specification? This might be adding too many features. It might be better to re-add at a later date if necessary.

File: qiskit/pulse/channels/pulse_specification.py [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+                self._controls == other._controls and \
+                self._measures == other._measures and \
+                self._acquires == other._acquires and \
+                self._mem_slots == other._mem_slots and \
+                self._reg_slots == other._reg_slots:
+            return True
+        return False
+
+    @property
+    def d(self) -> List[DriveChannel]:
```
Comment (taalexander) [inline qiskit/pulse/channels/pulse_specification.py]: Could these be something like `drives`, `memoryslots`, `qubits` etc? I think these names are too short.

File: test/python/compiler/test_assembler.py [hunk 6]
```diff
@@ -262,7 +262,7 @@ class TestPulseAssembler(QiskitTestCase):
     """Tests for assembling schedules to qobj."""
```
Comment (taalexander) [inline test/python/compiler/test_assembler.py]: The tests should be duplicated for now for `DeviceSpecification` until deprecation, although tests that use `DeviceSpecification` can just be switched to `PulseSpecification` as you have.

File: qiskit/pulse/channels/pulse_specification.py (hunk 4 above)
Comment (nkanazawa1989) [reply qiskit/pulse/channels/pulse_specification.py]: If we have some devices with the same topology, we don't need to recreate schedule. If we want to execute a schedule in parallel with multiple devices, we may want to check if the devices have the same topology. But I think such situation doesn't happen soon and we can remove equality check.

**Comment by nkanazawa1989** [comment]
@taalexander thank you for reviewing this. I updated the PR to align with your comments. unittests for `PulseSpecification` use `device.drives[0]` rather than `device.qubits[0].drive` because the later style implicitly includes qubit-channel mapping from device Hamiltonian. This should be tested separately in the new module `DeviceTopology` (I added this module and unittest for future extendability for device Hamiltonian).

Device Hamiltonian and n_drives/n_measures will be updated in following PRs after the spec is updated.

File: qiskit/pulse/channels/pulse_specification.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+"""
+from typing import List
+
+from .channels import AcquireChannel, MemorySlot, RegisterSlot
+from .pulse_channels import DriveChannel, ControlChannel, MeasureChannel
+from .qubit import Qubit
+from .device_topology import DeviceTopology
+
+
+class PulseSpecification:
```
[the remark this answers is not in the document]
Comment (nkanazawa1989) [reply qiskit/pulse/channels/pulse_specification.py]: That's good suggestion. I like `PulseChannelSpec`.

**Comment by nkanazawa1989** [comment]
@taalexander thank you for suggestions! the PR is updated.

File: qiskit/pulse/channels/pulse_channel_spec.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+        self._acquires = [AcquireChannel(idx, buffer) for idx in range(n_qubits)]
+        self._mem_slots = [MemorySlot(idx) for idx in range(n_qubits)]
+        self._reg_slots = [RegisterSlot(idx) for idx in range(n_registers)]
+
+        # create mapping information from channels
+        self._topology = SystemTopology(self._drives, self.controls,
+                                        self.measures, self.acquires)
+
+    @classmethod
+    def from_device(cls, backend):
```
Comment (nkanazawa1989) [inline qiskit/pulse/channels/pulse_channel_spec.py]: Do we change this to `from_system` ?

Comment (taalexander) [reply qiskit/pulse/channels/pulse_channel_spec.py]: Yes, but only in the new pulse channel spec and not the old DeviceSpecification.

File: qiskit/pulse/channels/pulse_channel_spec.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+from .pulse_channels import DriveChannel, ControlChannel, MeasureChannel
+from .qubit import Qubit
+from .system_topology import SystemTopology
+
+
+class PulseChannelSpec:
+    """A helper class to support assembling channel objects and their mapping to qubits.
+    This class can be initialized with two methods shown as below.
+
+    1. When you have BaseBackend object of the target quantum device.
```
Comment (taalexander) [inline qiskit/pulse/channels/pulse_channel_spec.py]:
```suggestion
    1. With the `BaseBackend` object of the target pulse backend:
```

File: qiskit/pulse/channels/pulse_channel_spec.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        ```python
+        system = PulseChannelSpec.from_device(backend)
+        ```
+
+    2. When you know the number of elements constituting the target quantum device.
```
Comment (taalexander) [inline qiskit/pulse/channels/pulse_channel_spec.py]:
```suggestion
    2. By specifying the number of pulse elements constituting the target quantum system:
```

**Comment by nkanazawa1989** [comment]
@taalexander thank you! I think every ~`device` is replaced by ~`system`.

File: qiskit/pulse/channels/pulse_channel_spec.py (hunk 10 above)
Comment (nkanazawa1989) [reply qiskit/pulse/channels/pulse_channel_spec.py]: I have a concern about using [quantum system](https://en.wikipedia.org/wiki/Quantum_system) to denote a system consisting of quantum chip + cryosystem + rte. This may remind us of a term in quantum physics. I think _quantum computing system_ sounds more suitable.

Comment (taalexander) [reply qiskit/pulse/channels/pulse_channel_spec.py]: yes, I agree. Your suggestion makes sense.

Comment (nkanazawa1989) [reply qiskit/pulse/channels/pulse_channel_spec.py]: thanks, I updated.

File: qiskit/pulse/channels/pulse_channel_spec.py [hunk 11]
```diff
@@ -91,9 +92,9 @@ def __init__(self,
                                         self.measures, self.acquires)
 
     @classmethod
-    def from_device(cls, backend):
+    def from_system(cls, backend):
```
Comment (taalexander) [inline qiskit/pulse/channels/pulse_channel_spec.py]: Do you think we should name this `from_backend` as this is what the method is doing?

Comment (nkanazawa1989) [reply qiskit/pulse/channels/pulse_channel_spec.py]: that's sounds reasonable.
